### ISSUE TITLE
Gradient of a real-valued function

### ISSUE THREAD

**Issue by mateuszbaran** [JuliaManifolds/Manifolds.jl#17]
We are definitely going to support optimization of real-valued functions defined on manifolds, so computing gradients of such functions is quite important. What we have now is:

- Manopt.jl seems to have some set of predefined gradients for certain functions but gradients need to be hand-written.
- pymanopt has some capability to use AD for gradients but I haven't looked into it that much.
- FunManifolds has good support for AD gradient on isometrically embedded manifolds (in R^n). I haven't yet thought about AD for MetricManifolds #12 .

There might also be some need for pushforwards, pullbacks etc.

I'll port some parts of FunManifolds's gradient capabilities here, do you have any specific things I should take into account?

**Comment by kellertuer** [comment]
Great topic. For now in Manopt.jl I am trying to help the user by providing necessary functions – but I'd love to extend that to AD capabilities of course.

As far as I read [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl) their main “trick” is to provide basic calculations like a chain rule as well as basic derivatives / gradients.

Using JacobiFields (well on symmetric manifolds available in closed form) we can at least provide a few derivatives/gradients, too (exp,log with respect to both base and argument, distance, norm of course), but we would have to define an extension to the Dual data type of ForwardDiff, i.e. to cover tangent vectors (I don't care much whether `TVector` or just `AbstractArray` but you might know I am in favour of the first).

**Comment by mateuszbaran** [comment]
I've checked pymanopt, they're just using normal AD frameworks and call `egrad2rgrad` to get the correct gradient. We could do a similar thing with ForwardDiff, at least initially. Ultimately, writing a custom, geometry-aware extension to Dual would be great.

For reference, Eric Forgy wrote a very interesting extension to standard AD: https://github.com/EricForgy/StochasticDiff.jl .

**Comment by kellertuer** [comment]
Yes, as a first approach the way pymanopt introduces AD is just fine – though my idea would be to use a more general function name `project` (similar to projecting onto a manifold, just onto a tangent space) instead of the name they use, since one could also project any tangent vector from the embedding onto a tangent plane.

In the long run I was exactly thinking of an extension to standard AD that is geometry aware.

**Comment by mateuszbaran** [comment]
The thing about the `egrad2rgrad` is that for some manifolds they are storing the tangent vector differently (it's not a simple projection). BTW, we already have `project_tangent` for sphere.

**Comment by kellertuer** [comment]
Yes, I know, for example on `SO(n)` I do so, too – that's one of the points where I decided to use types for points and vectors.
Ah yes, of course we have.

**Comment by mateuszbaran** [comment]
I'm thinking about it now and I wonder in what situations one would use `project_tangent` that is different from `egrad2rgrad`. Just as a simple way to do approximate vector transport? Or something else?

**Comment by kellertuer** [comment]
Yes, for example when approximating vector transport, or whenever you end up in an embedded tangent vector setting and want to come back, but vector transport is the main origin of the more general name, yes.

**Comment by sdaxen** [comment]
> Ultimately, writing a custom, geometry-aware extension to Dual would be great. 

> In the long run I was exactly thinking of an extension to standard AD that is geometry aware.

A relevant reference for geometry-aware AD: https://arxiv.org/abs/1812.11592

**Comment by mateuszbaran** [comment]
I think that's more or less right regarding gradients and Jacobians should be more or less just as simple. Real fun begins when you start considering vector fields, covariant derivatives and Hessians though, there are Christoffel symbols everywhere.

**Comment by kellertuer** [comment]
Just to update this – I am currently reading through the book by Nicolas and I think I understood the reason for the `egrad2rgrad` function.

1) If you have a manifold, where the metric on each tangent space is obtained by restricting the metric of the embedding to the tangent space (i.e. for the sphere or what we refer to as `IsometricEmbedding`), then the function is just the projection and maybe a change in representation (for example what `project` will do on fixed rank matrices anyways).
2) If that is not the case, you also have to take this into account, basically “reverse engineer” the differential from the gradient and find the other Riesz representer (with respect to the other metric used for the manifold). This should still be doable, i.e. working through the `egrad2rgrad`s or the theory of a manifold to see how that works.

I could start this either here or in a separate repository – I am not yet sure how to call this conversion in (2) (the case (1) would just pass the same function over to project). At least one of my goals next is to better understand how we can employ more AD here :) (also write tutorials for the existing methods then of course).

**Comment by mateuszbaran** [comment]
Why couldn't `project` do the conversion in case (2)?

**Comment by kellertuer** [comment]
I have to read more about this, before I can definetly answer that. For the implementation this would be nice, for sure – just for the user I am not so sure whether also doing this in `project` would maybe be a misuse of that function (or its name to be precise).

**Comment by mateuszbaran** [comment]
OK, so maybe let's work out an example where `project` and `egrad2rgrad` should do something different and then figure out the changes to out interface that need to be done.

**Comment by turnersr** [comment]
I was looking at another resource: https://files.cercomp.ufg.br/weby/up/328/o/iusfen3.pdf . The section I am referring to is where they construct the covariant derivative. Instead of projecting the output as in projected gradient descent, this method appears to create a covariant derivative by projecting the normal gradient onto the manifolds tangent vectors. 

So there appear to be 2 overall ways and under those several distinct ways to optimize over manifolds:

1. Constrain the output of the optimization 

- Project the output of each iteration of descent as in project gradient descent 
- Form a new gradient by projecting the classical gradient on to the tangent space of the surface 
- Create an intrinsically defined gradient, for example, using spherical coordinates, to define the gradient on say the sphere. 

2. Constrain the loss function using information about the surface 

- Create a loss function that is sensitive to items in the manifold. For example, you can define a distance between any matrix an element of the SO(n).  As in https://github.com/geomstats/applications/tree/master/se3_pose_estimation 

What are the different trade-offs of these approaches? Why would one prefer to use one over the other? 

Hope this helps.

**Comment by kellertuer** [comment]
Hi and thank you for your input.

Concerning your first point: projection is exactly what I want to investigate generically, i.e. for any manifold, but I am not yet sure whether this works on any manifold. For the sphere, sure that's really just classical projection, but keep in mind that for example for symmetric positive definite matrices (represented as matrices) there is no projection for points – and I am looking into whether a projection is always available for tangents and to what extend the metric has to be taken into account when the manifold is not isometrically embedded. Your reference (I am aware of works by Ferreira and Nemeth) is (just) for the sphere; on spheres projection the classical gradient from the embedding onto the tangent space yields the Riemannian gradient and we have `project!(M,Y,p,X)` for that :)

For the second point: Well our `distance(Max,y)` is generically exactly that and for the distance as a specific example, the gradient is known, see https://manoptjl.org/stable/functions/gradients.html#Manopt.grad_distance but
* it requires a closed form logarithmic map, which might not be available for the manifold in mind (Stiefel mit the Euclidean metric for example)
* it is just _one_ function and while it might be good to collect a few gradients that are “nice” to compute (see the page I just linked) a more generic automatic way would be preferable (maybe using the gradients/differentials that are available as ingredients for sure).

So in short the second approach is not generic enough, unless someone (not me) collects a large large library of gradients – and even then a generic automatic way (AD) would definitely be less to implement and faster I hink

For the first approach – that might be the forward mode, as soon as we are sure that generically works, that's exactly what I want to look at next.

**Comment by kellertuer** [comment]
I think I now understand what the `egrad2rgrad` function does: Let M be an embedded manifold (in say E). Then there is three cases that illustrate what this function does in Manopt (Matlab)

1) For the sphere, we have an isometric embedding, i.e. the metric of the tangent space at `p` is obtained by restricting the metric from the embedding to the tangent space. Then the transformation is just a projection  https://github.com/NicolasBoumal/manopt/blob/3774b069666a3a1a01a33e10f53062e33495e59e/manopt/manifolds/sphere/spherefactory.m#L122 (similarly for example Stiefel, rotations,...)
2) if it is not isometrically embedded, we have to account for the “change of metric” when the metric for the manifold is another metric, maybe even depending on `p` (in manopt/Matlab called `X`), then this change in metric has to be accounted for, basically by multiplying with G^{ij} (the inverse matrix of the metric). See for example the symmetric positive definite matrices at https://github.com/NicolasBoumal/manopt/blob/3774b069666a3a1a01a33e10f53062e33495e59e/manopt/manifolds/symfixedrank/sympositivedefinitefactory.m#L102 which is in my opinion two steps (a) project onto the tangent space (b) change inner product representation (another example is the hyperbolic space in the default hyperboloid representation, see https://github.com/NicolasBoumal/manopt/blob/3774b069666a3a1a01a33e10f53062e33495e59e/manopt/manifolds/hyperbolic/hyperbolicfactory.m#L166)
3) Sometimes, even more, the representation changes. This can best be seen for the fixed rank matrices, that we currently use, i.e. in their 3-factor-representation. The example is https://github.com/NicolasBoumal/manopt/blob/3774b069666a3a1a01a33e10f53062e33495e59e/manopt/manifolds/fixedrank/fixedrankfactory_3factors.m#L80-L85 – there I think Manopt is even a little inconsistent, because the incoming euclidean gradient is already given as an element with three components. I would have expected (a) first a projection onto the tangent space (which is the identity for fixed rank) (b) the change of representation from a single matrix to the 3-factor representation and (c) the change in the inner product. In this case only (a) is performed but already in the form that I would have expected to be only produced in (b).

To summarise
a) projection of the gradient from the embedding onto the tangent space
b) change of representation
c) change of inner product

As far as I can see, our project currently does a) and b) if b is necessary, but for example depending on the input we do it with or without b) here https://github.com/JuliaManifolds/Manifolds.jl/blob/b0459de13f7818a3cae16b94bd145da3104fcc04/src/manifolds/FixedRankMatrices.jl#L312-L323

But for c) I am not sure, should we include this in the `project!(M, Y, p, X)`? Currently (cf. https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/hyperbolic.html#ManifoldsBase.project-Tuple{Hyperbolic,Any,Any} ) we do not do that in the projection, we just project onto the tangent space (using the inner product in the embedding). If we include c) consistently, our variant of `egrad2rgrad` would just be `project!` – for me it would be more like a function `change_metric(::Manifold, ::Metric1, ::Metric2, p, X`, and then

`convert_gradient!(M, Y, p, X)` would convert the Euclidean gradient `X` into a Riemannian `Y` (and basically do `project` and `change_metric`). What do you think?

side note: Similar thoughts we might have to do with `ehess2rhess`, but that's a next issue/task.

**Comment by turnersr** [comment]
@kellertuer , awesome! I think one main difference between 1) projecting the output every step of descent and 2) forming the covariant derivative and moving in the direction given by the exponential map, is that 1) is optimizing in the ambient Euclidean vector space and then projecting onto the tangent space of the manifold whereas 2) is optimizing in a way that is intrinsic to the manifold without regards to the ambient space in which it could be in embedded in. 

I think both approaches are valuable no? But I am not really sure when to pick one vs the other. For example, if you do project gradient descent on a Klien bottle that seems like it would not work well. However, if you simply scale the output of gradient descent by the norm that seems ok and just as good as using a gradient instinct to the manifold.  

There is another library that has a projection map from the ambient space to the manifold of SPD matrices: https://github.com/master/tensorflow-manopt/blob/master/tensorflow_manopt/manifolds/symmetric_positive.py#L37 

does this answer the question is if there is a point projector for symmetric positive definite matrices?

**Comment by kellertuer** [comment]
I think you have maybe misread something, we _never_ optimise in ambient space, we always work intrinsically.
The only thing we aim to do in this issue (model and implement) is to use AD to compute a Riemannian gradient by computing an Euclidean gradient in the embedding and (if that is possible) compute the Riemannian one.

Keep in mind that there are even manifolds that are not embedded manifolds, then this approach would not work. For those I would prefer an _inttrinsic_ way to even do AD – but I have not seen that yet.

For the SPDs we have a very good example: If you do an optimization step in the embedding (just + on matrices) you might end up with a matrix that is not Spd – but the set of SPD matrices in the set of matrices is _open_ (!). You can not project on an open set! For SPDs all eigenvalue have to be positive! So I am sorry to report on the example you give, but having zero eigenvalues means, that the matrix is only positive semidefinite! These do not belong to the manifold of SPDs. The projection you linked is not a projection onto SPDs. There is not projection of points onto the SPDs!

So there also never was e question whether there is a projection, since there really _is not_.

**Comment by turnersr** [comment]
ooooooh! thank you so much for your clarifications. This makes sense. I wasn't aware you all didn't want to optimize in the ambient space and then project onto the tangent space of the feasibility manifold. I don't know enough to have a strong opinion here. I wonder when one approach is picked over the other. 

I see now. Your example of SPDs is true! I wonder what the tensorflow_manopt library is doing then. I assumed it was somehow justified, but now I have my doubts. I will have to study what this so-called "projection" is actually doing. 

I wonder how people deal with manifolds that can't be embedded. Your intrinsic approach of computing the covariant derivative and moving in the proper direction with retraction or exponential map makes sense for many embeddable manifolds. Now I am very curious about optimizing over PSL(2,R), for example.

**Comment by kellertuer** [comment]
If you have an embedded manifold (like the sphere) you can use the a metric in the embedding (for the sphere really the canonical usual metric in R^n+1) to project onto the manifold. Similarly to project vectors onto tangent spaces. But this only works for embedded manifolds, if they are not open sets.

I am not yet aware, what PSL(2,R) is, but feel free to ask details in a new issue here or on discourse. It would lead to far away from the topic here, computing the gradient with AD.

**Comment by mateuszbaran** [comment]
> 3\. Sometimes, even more, the representation changes. This can best be seen for the fixed rank matrices, that we currently use, i.e. in their 3-factor-representation. The example is https://github.com/NicolasBoumal/manopt/blob/3774b069666a3a1a01a33e10f53062e33495e59e/manopt/manifolds/fixedrank/fixedrankfactory_3factors.m#L80-L85 – there I think Manopt is even a little inconsistent, because the incoming euclidean gradient is already given as an element with three components. I would have expected (a) first a projection onto the tangent space (which is the identity for fixed rank) (b) the change of representation from a single matrix to the 3-factor representation and (c) the change in the inner product. In this case only (a) is performed but already in the form that I would have expected to be only produced in (b).

That's an interesting point, that seems to consider fixed-rank `n x m` matrices to be embedded in a product manifold of two Stiefel manifolds and I think the third is something like positive r-element vector embedded in R^{r \times r}? See https://arxiv.org/pdf/1209.0430.pdf , Fig. 1. Would it make sense to model the 3-factor representation of fixed-rank matrices using that product manifold?

> But for c) I am not sure, should we include this in the `project!(M, Y, p, X)`? Currently (cf. [https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/hyperbolic.html#ManifoldsBase.project-Tuple{Hyperbolic,Any,Any}](https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/hyperbolic.html#ManifoldsBase.project-Tuple%7BHyperbolic,Any,Any%7D) ) we do not do that in the projection, we just project onto the tangent space (using the inner product in the embedding). If we include c) consistently, our variant of `egrad2rgrad` would just be `project!` – for me it would be more like a function `change_metric(::Manifold, ::Metric1, ::Metric2, p, X`, and then
> 
> `convert_gradient!(M, Y, p, X)` would convert the Euclidean gradient `X` into a Riemannian `Y` (and basically do `project` and `change_metric`). What do you think?

Well, I that `change_metric` function feels really weird. Currently We assume that a single manifold has a single metric, so if anything, a `change_metric` function would actually be a map between two different tangent bundles. We have a weird situation here because the way we use `AbstractEmbeddedManifold` implicitly assumes `embed` is generally a no-op, which means we have to "fix" the tangent vector somewhere else.

By non-isometric embedding here we mean that embedding's `inner` after no-op embedding is different that the metric of the embedded manifold but maybe that's totally fine and we should just assume that `embed` always performs an isometric embedding of tangent spaces? I'm not sure here, that's just an idea to consider.

> Keep in mind that there are even manifolds that are not embedded manifolds, then this approach would not work. For those I would prefer an _inttrinsic_ way to even do AD – but I have not seen that yet.

I think all intrinsic AD would require is an intrinsic way to do pushforwards and pullbacks, as AD is all about pushforwards and pullbacks. I've done cotangent vectors in #325 with this in mind. I'd be happy to discuss this further if you want to work on that.

> I am not yet aware, what PSL(2,R) is, but feel free to ask details in a new issue here or on discourse. It would lead to far away from the topic here, computing the gradient with AD.

That's probably projective special linear group?

**Comment by kellertuer** [comment]
> That's an interesting point, that seems to consider fixed-rank `n x m` matrices to be embedded in a product manifold of two Stiefel manifolds and I think the third is something like positive r-element vector embedded in R^{r \times r}? See https://arxiv.org/pdf/1209.0430.pdf , Fig. 1. Would it make sense to model the 3-factor representation of fixed-rank matrices using that product manifold?

I think that is also what we actually do, if you interpret the U,V matrices of the SVD in their reduced (i.e. non-square) form, so we are actually doing something like that, we just don't write it as a product manifold nor do we use the `ProductRepr`. I am not sure whether changing this would have much of an effect?

> 
> > But for c) I am not sure, should we include this in the `project!(M, Y, p, X)`? Currently (cf. [https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/hyperbolic.html#ManifoldsBase.project-Tuple{Hyperbolic,Any,Any}](https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/hyperbolic.html#ManifoldsBase.project-Tuple%7BHyperbolic,Any,Any%7D) ) we do not do that in the projection, we just project onto the tangent space (using the inner product in the embedding). If we include c) consistently, our variant of `egrad2rgrad` would just be `project!` – for me it would be more like a function `change_metric(::Manifold, ::Metric1, ::Metric2, p, X`, and then
> > `convert_gradient!(M, Y, p, X)` would convert the Euclidean gradient `X` into a Riemannian `Y` (and basically do `project` and `change_metric`). What do you think?
> 
> Well, I that `change_metric` function feels really weird. Currently We assume that a single manifold has a single metric, so if anything, a `change_metric` function would actually be a map between two different tangent bundles. We have a weird situation here because the way we use `AbstractEmbeddedManifold` implicitly assumes `embed` is generally a no-op, which means we have to "fix" the tangent vector somewhere else.

I am also not yet sure, since I did not yet think so much about `embed!` for non-isometrically embedded manifolds (like SPD), but you are right, there is no adaption to a change of the metric in these cases. I am just not yet sure how to model it instead of with a `change_metric` function.

Classically a `change_metric(M1,M2,p,X)` would have two `MetricManifolds` decorating the same manifold (say the SPDs) and you would change `X` such that the result `Y` is of same length (in metric 2) as `X` (in metric 1). If M1 and M2 are not based on the same manifold but M2 is a submanifold and `X` is from its tangent space, this change of metric is possible, too, for sure.

> By non-isometric embedding here we mean that embedding's `inner` after no-op embedding is different that the metric of the embedded manifold but maybe that's totally fine and we should just assume that `embed` always performs an isometric embedding of tangent spaces? I'm not sure here, that's just an idea to consider.

I think we implicitly assumed that or we just embedded the vector as is (not taking into account that embedding might change its length since the metric changes). And I even think embed should just change the representation of the tangent vector to a tangent vector in the embedding (so more thinking of both as sets), as we do now (so it is mostly a no-op).

> > Keep in mind that there are even manifolds that are not embedded manifolds, then this approach would not work. For those I would prefer an _inttrinsic_ way to even do AD – but I have not seen that yet.
> 
> I think all intrinsic AD would require is an intrinsic way to do pushforwards and pullbacks, as AD is all about pushforwards and pullbacks. I've done cotangent vectors in #325 with this in mind. I'd be happy to discuss this further if you want to work on that.

exactly, that's what I would like to improve here somewhen, but maybe lets first start with the embedding approach, like geomstats and Manopt and pymanopt do, too. Maybe I keep all this also in mind for a master student. Sounds nice.

> 
> > I am not yet aware, what PSL(2,R) is, but feel free to ask details in a new issue here or on discourse. It would lead to far away from the topic here, computing the gradient with AD.
> 
> That's probably projective special linear group?

Thanks, I did not think of that one.

**Comment by kellertuer** [comment]
Or to summarise: For me `embed` is on the same level as `is_manifold_point` or `is_tangent_vector`, namely a set/representation level function (sure project is not...).

**Comment by turnersr** [comment]
`If you have an embedded manifold (like the sphere) you can use the a metric in the embedding (for the sphere really the canonical usual metric in R^n+1) to project onto the manifold. Similarly to project vectors onto tangent spaces. But it only works for embedded manifolds, if they are not open sets.`

When you project onto an embedded manifold during optimization how do you currently account for the change in metric? Is that taken care of by the covariant derivative followed by the retraction approach? 

I will make new issue for PSL(2,R). 

This is a really awesome project. Thanks for your help.

**Comment by mateuszbaran** [comment]
> I think that is also what we actually do, if you interpret the U,V matrices of the SVD in their reduced (i.e. non-square) form, so we are actually doing something like that, we just don't write it as a product manifold nor do we use the `ProductRepr`. I am not sure whether changing this would have much of an effect?

I've checked our code and indeed there seems to be not much reason to change fixed rank manifold this way.

> Classically a `change_metric(M1,M2,p,X)` would have two `MetricManifolds` decorating the same manifold (say the SPDs) and you would change `X` such that the result `Y` is of same length (in metric 2) as `X` (in metric 1). If M1 and M2 are not based on the same manifold but M2 is a submanifold and `X` is from its tangent space, this change of metric is possible, too, for sure.

That's just pushforward of "identity" function between M1 and M2, right? Maybe we could model it as a special case of a pushforward?

> I think we implicitly assumed that or we just embedded the vector as is (not taking into account that embedding might change its length since the metric changes). And I even think embed should just change the representation of the tangent vector to a tangent vector in the embedding (so more thinking of both as sets), as we do now (so it is mostly a no-op).

> Or to summarise: For me `embed` is on the same level as `is_manifold_point` or `is_tangent_vector`, namely a set/representation level function (sure project is not...).

I just realized we could also have tangent vector embeddings as pushforwards of point embeddings. Now we need to decide whether projection is an extension of the inverse of our current vector `embed` or of the inverse of that pushforward of point `embed`. We probably need both but we need a nice API.

**Comment by mateuszbaran** [comment]
> exactly, that's what I would like to improve here somewhen, but maybe lets first start with the embedding approach, like geomstats and Manopt and pymanopt do, too. Maybe I keep all this also in mind for a master student. Sounds nice.

Sure, starting with an embedding approach is exactly what I did in https://github.com/JuliaManifolds/Manifolds.jl/pull/195 . There are many directions that could be taken from here, for example integration with Manopt.jl, more thorough testing, adding pushforwards and pullbacks, etc.

> When you project onto an embedded manifold during optimization how do you currently account for the change in metric? Is that taken care of by the covariant derivative followed by the retraction approach? Is this what you call the "embedding approach" for intrinisic optimization like in geomstats, Manopt and pymanopt?
> 
> Doesn't the metric change happen even for the sphere's case? Not sure what this implicit assumption is. Maybe I missing something big, but without accounting for the change it would be hard to optimize, no? It would be nice to get around it, if you need help, if it is blocking on optimizing over non embeddable manifolds. Trying to catch up. This discussion as been helpful.

IIUC no projections are actually performed during optimization :slightly_smiling_face: . I've found only one relevant projection in Manopt.jl in `QuasiNewtonLimitedMemoryDirectionUpdate{InverseBFGS}` but I guess that update rule only works for certain manifolds? I'm sure @kellertuer  will be able to explain it better than me :slightly_smiling_face: .

**Comment by kellertuer** [comment]
> > Doesn't the metric change happen even for the sphere's case? Not sure what this implicit assumption is. Maybe I missing something big, but without accounting for the change it would be hard to optimize, no? It would be nice to get around it, if you need help, if it is blocking on optimizing over non embeddable manifolds. Trying to catch up. This discussion as been helpful.
> 
> IIUC no projections are actually performed during optimization 🙂 . I've found only one relevant projection in Manopt.jl in `QuasiNewtonLimitedMemoryDirectionUpdate{InverseBFGS}` but I guess that update rule only works for certain manifolds? I'm sure @kellertuer will be able to explain it better than me 🙂 .

So first o the sphere. The metric of a tangent space TpM, M = Sphere(n) as an embedded manifold obtains its metric by restricting the euclidean metric to the tangent space (i.e. you take the same metric – just x^Ty – but only elements from the tangent space are allowed as input. So there is no change.

In Manopt.jl (Manopt/pymanopt) all operations are always intrinsic, no projections required. With the small caveat, that the limited memory quasi newton with inverse BFGS was not stable when my student had finished the implementation, so for now – but just for stability – the projection is in there. I did not yet have the time to investigate, where the small error might be. But in principle, all methods (gradient descent, cyclic proximal point etc pp) do not work on the embedding with projection but just intrinsically without leaving the manifold/tangent space.

**Comment by kellertuer** [comment]
> > exactly, that's what I would like to improve here somewhen, but maybe lets first start with the embedding approach, like geomstats and Manopt and pymanopt do, too. Maybe I keep all this also in mind for a master student. Sounds nice.
> 
> Sure, starting with an embedding approach is exactly what I did in #195 . There are many directions that could be taken from here, for example integration with Manopt.jl, more thorough testing, adding pushforwards and pullbacks, etc.

Yes, that's why I started continuing here do see a few things, and look where and how it can be used; I can surely do a tutorial how to use your parts (at least for the sphere for example) in Manopt.jl – it just does not yet work for _all_ manifolds due to the missing change in representation(case 3b above) or change of metric (3c above). But for cases where neither of that is needed (again, the sphere) your parts from #195 should work fine :)

**Comment by kellertuer** [comment]
Coming back to the topic discussed here, I think we have three possible situations when converting A Euclidean Gradient (maybe from AD) to a Riemannian one

(1) a projection is required, like for the sphere - but nothing more
(2) a projection is required and we have to switch the metric by multiplying with the (inverse?) of the metric tensor (in some representation); this would be the case for the SPDs, since the tangent space is all symmetric matrices but we also have to adapt the metric - I think this should be part of project as well?
(3) For Lie groups we additionally have to move to the Lie algebra; should this be part of project as well since it is just a change of representation but it yields the representation the manifold indeed has?

If all three are within project, then indeed turning a Euclidean gradient into a Riemannian one would just be project. We should then maybe add that :)

**Comment by mateuszbaran** [comment]
I think we should separate projection onto the tangent space and representation change. There are valid use cases for projection that assume no representation change happens, for example correcting numerical errors (even Manopt.jl does that) or IIRC some Hessian calculation methods. Manopt and pymanopt also differentiate between tangent space projection and `egrad2rgrad`. I don't really see much benefit in merging these two.

**Comment by kellertuer** [comment]
but when I do these projections, the metric currently does not change.

Ok then we should not merge them, but keep all three scenarios atop in mind. Then the question is: How to call this function?
It takes a tangent vector,
It projects, adapts to the new metric so the norm does not change (wrt the new metric, its even a little more, when multiplying with the metric tensor) – and eventually switches to the Lie algebra, and you are right until now (and maybe we should keep it that way) project does not do steps 2&3.

we should keep in mind that we want to do something similar with (1,1) tensors for the hessian, too, but I do not have a good idea for a name.

**Comment by mateuszbaran** [comment]
This is complicated... also note that another useful and related thing is differential of point projection.

Maybe `project_ambient` would only do (1), `project` would do (1) + (3) + maybe (2) depending on the projection type? I don't think we need (3) without (1), and (2) could be its own function. If I understand correctly (2) would be the differential of identity on the manifold?

**Comment by kellertuer** [comment]
The differential does not “see” the metric, the differential of the identity is the identity :D. But the Riesz representer of this differential that characterises roughly what you mean.

I think introducing
* `project_ambient` declaring it does not change the metric after projection is a good idea (and maybe also what Manopt.jl should use then)
* I think `project` should so 1-3; (2) if the metric changes (otherwise there is nothing to do if it is isometrically embedded) and for sure (3) is never needed along I think.
* sure (2) (for the sphere being the identity) could be its own function something like `change_metric`.

**Comment by mateuszbaran** [comment]
Right, of course, differential does not depend on the metric at all. Could you elaborate a bit on what that Riesz representer thing? I don't see how it would work in this case.

**Comment by kellertuer** [comment]
For the differential that is “just” the gradient (indeed one way to define the gradient: The unique tangent vector Y for which

Df[X] = <X, Y>

holds (for all X). Then Y depends on the metric for sure.

**Comment by mateuszbaran** [comment]
And what is `f`?

**Comment by kellertuer** [comment]
ah sorry that was for the gradient of an arbitrary f, for you the identity...ah! Maybe it is already too late for my brain, for sure this equation only works for f mapping into the reals. Then I am not sure how your identity differential should help determining the metric change – I only know how to obtain the metric tensor (i.e. in coordinates as a matrix) and “compensate” for it (i.e. multiply with its inverse).

So what did you mean with the differential of the identity, when that does not “see” changes of the metric?

**Comment by mateuszbaran** [comment]
Differential of identity may not be isometric when change of metric happens. The problem with multiplying by the inverse metric tensor is that it's rather specific to a single scenario... if we wanted `f:T_pM -> T_pN` such that `<X, Y>_M == <f(X), f(Y)>_N` (when M, N are the same manifold with different metrics) then I think `f`  be different? Let's have `f(X) = BX`, then `X^T g_M Y = (B X)^T g_N B Y ` so we need `B` such that `g_M  = B^T g_N B`?

**Comment by kellertuer** [comment]
Ah, I see. Sure, I did not say that multiplying with the inverse of the metric tensor is directly _code_ you can implement ;) But with B (and not the identity) this looks like a nice idea.

**Comment by kellertuer** [comment]
I think I now understood how you mean the differential of the identity. I will start with a branch that first collects maybe a little more details on how I think we can model the change of the metric, since this seems the most-interesting part of the three steps

' the projection we already have
* the change of representation is either conversion into a different `ManifoldPoint` type (for example for Fixed Rank) or a parallel transport into the identity (for Lie groups)

**Comment by kellertuer** [comment]
The docs in the new PR explain, that it is indeed close to the differential of the identity what you meant, just that the differential is independent of the metric – but your explanation helped me writing the current (hopefully right) approach there.

**Comment by mateuszbaran** [comment]
Hm, nice, I'll check your PR.
